File: serverless/express-based-api-with-dynamodb/entities/responses.js
// Load the AWS SDK for JS
var AWS = require("aws-sdk");
AWS.config.update({region: 'us-east-1'});

// Load the UUID library
const uuidv4 = require('uuid/v4');

// Create the DynamoDB Document Client
var dynamodb = new AWS.DynamoDB.DocumentClient();
var tableName = process.env.DYNAMODB_TABLE

module.exports.create = async function(body) {
    const survey_id = body['survey_id']
    const response_id = uuidv4()
    const response_data = body['response_data']
    const putParams = {
        TableName: tableName,
        Item: {
            'pk': 'RESPONSE#' + response_id,
            'sk': 'SURVEY#' + survey_id,
            'response_data': response_data
        }
    }
    try {
        await dynamodb.put(putParams).promise()
        return {
            "response_id": response_id
        }
    } catch (error) {
        console.log(error)
        throw new Error(error)
    }
}

module.exports.get = async function(response_id) {
    const queryParams = {
        TableName: tableName,
        KeyConditionExpression: 'pk = :pk',
        ExpressionAttributeValues: {
            ':pk': 'RESPONSE#' + response_id
        }
    }
    try {
        const queryResult = await dynamodb.query(queryParams).promise()
        console.log(queryResult)
        return queryResult['Items'][0]
    } catch (error) {
        console.log(error)
        throw new Error(error)
    }
}


module.exports.getAll = async function(survey_id) {
    const queryParams = {
        TableName: tableName,
        IndexName: 'sk-pk-index',
        KeyConditionExpression: 'sk = :sk AND begins_with ( pk , :pk )',
        ExpressionAttributeValues: {
            ':pk': 'RESPONSE#',
            ':sk': 'SURVEY#' + survey_id
        }
    }
    try {
        const queryResult = await dynamodb.query(queryParams).promise()
        console.log(queryResult)
        return queryResult['Items']
    } catch (error) {
        console.log(error)
        throw new Error(error)
    }
}

File: serverless/flask-based-api/entities/responses.py
import boto3
import os
import json
import uuid

from boto3.dynamodb.conditions import Key

dynamodb = boto3.resource('dynamodb')

TABLE_NAME = os.environ['DYNAMODB_TABLE']
table = dynamodb.Table(TABLE_NAME)


def create(body):
    survey_id = body['survey_id']
    response_id = str(uuid.uuid4())
    response_data = body['response_data']
    table.put_item(
        Item={
            'pk': 'RESPONSE#' + response_id,
            'sk': 'SURVEY#' + survey_id,
            'response_data': response_data
        }
    )
    return json.dumps({"response_id": response_id})


def get(response_id):
    index_pk = Key('pk').eq('RESPONSE#' + response_id)
    response = table.query(
        KeyConditionExpression=index_pk
    )
    return json.dumps(response['Items'][0])


def get_all(survey_id):
    index_pk = Key('sk').eq('SURVEY#' + survey_id)
    index_sk = Key('pk').begins_with('RESPONSE#')
    expression = index_pk & index_sk
    response = table.query(
        IndexName='sk-pk-index',
        KeyConditionExpression=expression
    )
    return json.dumps(response['Items'])
